### PR TITLE
slack plugin initial version

### PR DESCRIPTION
**Pull request by lingpri**
Ability to send a slack message using graylog notification system.

## Description
the slack notification needs to be upgraded.

## Motivation and Context
https://github.com/Graylog2/graylog-plugin-enterprise/issues/1217

## How Has This Been Tested?
slack_notification_local_testing

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by kyleknighted** [comment]
Could you implement https://graylog2.github.io/frontend-documentation/#!/ColorPicker with the ability to manually add the hex color with a preview of the color?

**Comment by waab76** [comment]
Failed while attempting to send a test event during setup:

Screen Shot 2020-08-31 at 12 29 24 PM

**Comment by waab76** [comment]
Unable to save Slack notification:

Screen Shot 2020-08-31 at 12 31 05 PM

**Review by bernd — CHANGES_REQUESTED** [review]
See inline comments.

Another question: what's the reason for putting this into the server core instead of the integrations plugin?

File: graylog2-server/src/main/java/org/graylog/events/notifications/types/SlackClient.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.net.*;
```
[the remark this answers is not in the document]
Comment (lingpri) [reply graylog2-server/src/main/java/org/graylog/events/notifications/types/SlackClient.java:29]: @bernd the components for `EmailNotification` exists in the project `graylog2-server`, hence i raised the PR for slack-plugin in `graylog2-server` repository.

File: graylog2-server/src/main/java/org/graylog/events/notifications/types/SlackClient.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		try {
+			url = new URL(webhookUrl);
+		} catch (MalformedURLException e) {
+			throw new SlackClientException("Error while constructing webhook URL.", e);
+		}
+
+		final HttpURLConnection conn;
+		try {
+			if (!StringUtils.isEmpty(proxyURL)) {
+				final URI proxyUri = new URI(proxyURL);
```
[the remark this answers is not in the document]
Comment (lingpri) [reply graylog2-server/src/main/java/org/graylog/events/notifications/types/SlackClient.java:55]: I will make the changes.

File: graylog2-server/src/main/java/org/graylog/events/notifications/types/SlackEventNotification.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+	private Map<String, Object> getCustomMessageModel(EventNotificationContext ctx, SlackEventNotificationConfig config, List<MessageSummary> backlog) {
+		Optional<EventDefinitionDto> definitionDto = ctx.eventDefinition();
+
+		List<StreamModelData> streams = streamService.loadByIds(ctx.event().sourceStreams())
+				.stream()
+				.map(stream -> buildStreamWithUrl(stream, ctx, config))
+				.collect(Collectors.toList());
+
+		CustomMessageModelData modelData = CustomMessageModelData.builder()
```
[the remark this answers is not in the document]
Comment (lingpri) [reply graylog2-server/src/main/java/org/graylog/events/notifications/types/SlackEventNotification.java:194]: Yeah, I will look into refactoring the code.

File: graylog2-server/src/main/java/org/graylog/events/notifications/types/SlackClient.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog/events/notifications/types/SlackClient.java]: @lingpri Please move this PR to the graylog-plugin-integrations repository.

**Comment by lingpri** [comment]
> Could you implement https://graylog2.github.io/frontend-documentation/#!/ColorPicker with the ability to manually add the hex color with a preview of the color?

@kyleknighted - I forgot, why I need a color picker for the slack plugin. could you please remind me. Also, @bernd has suggested that I move the code to the `graylog-plugin-integrations` project. 

> Could you implement https://graylog2.github.io/frontend-documentation/#!/ColorPicker with the ability to manually add the hex color with a preview of the color?

@kyleknighted - is this for the `notification-color` element in the form?

```
<Input id="notification-color"
               name="color"
               label="Custom Message Color"
               type="text"
               bsStyle={validation.errors.color ? 'error' : null}
               help={lodash.get(validation, 'errors.color[0]', 'Color to use for Slack message')}
               value={config.color || ''}
               onChange={this.handleChange}
               required />
```

**Comment by kyleknighted** [comment]
@lingpri 
>  I forgot, why I need a color picker for the slack plugin. could you please remind me. 

I do not know the requirements for this plugin, so I am not sure why you need a color. Perhaps @bud1979 or @waab76 know the requirements and why you need color.

>  is this for the notification-color element in the form?

Yes, if the requirements ask for a user provided color, please use the the ColorPicker component.

**Comment by lingpri** [comment]
1. https://github.com/Graylog2/graylog-plugin-integrations/pull/618
2. https://github.com/Graylog2/graylog-plugin-integrations/pull/595

**Comment by lingpri** [comment]
the initial version is in plugin integrations branch .

**Pull request closed without merging**